src/protocol/sort/secureapplyinv.rs

```diff
@@ -0,0 +1,171 @@
+    input: &'a mut Vec<Replicated<F>>,
+    permutation: &'a mut Vec<Replicated<F>>,
+}
+
+impl<'a, F: Field> SecureApplyInv<'a, F> {
+    #[allow(dead_code)]
+    pub fn new(input: &'a mut Vec<Replicated<F>>, permutation: &'a mut Vec<Replicated<F>>) -> Self {
+        Self { input, permutation }
+    }
+    /// In a nutshell, this algorithm applies a permutation on the inputs. The permutation is not available in clear and
+    /// we do not want to reveal the actual permutation to any of the helpers while applying it.
+    /// Steps
+    /// ![Secure Apply Inv steps][secureapplyinv]
+    /// 1. Generate a permutation
+    /// 2. Secret shared permutation is shuffled with a permutation
+    /// 3. Secret shared value is shuffled using the same permutation as 1
+    /// 4. The permutation is revealed
+    /// 5. All helpers call `apply` to apply the permutation locally.
+    #[allow(dead_code)]
+    pub async fn execute<N: Network>(
+        &mut self,
+        ctx: &ProtocolContext<'_, N>,
+    ) -> Result<(), BoxError> {
+        // 1. Generate permutation
+        let permutations = generate_random_permutation(self.input.len(), &ctx.prss());
+        // 2. Shuffle permutations
+        let mut shuffle_permutation = Shuffle::new(self.permutation);
+        let h0_future = shuffle_permutation.execute(ctx.narrow(&ShufflePermutation), &permutations);
+
+        // 3. Shuffle inputs
+        let mut shuffle_input = Shuffle::new(self.input);
+        let h1_future = shuffle_input.execute(ctx.narrow(&ShuffleInputs), &permutations);
+
+        try_join!(h0_future, h1_future).unwrap();
+
+        // 4. Reveal permutation to all helpers
+        let permutation = self
+            .reveal_permutations(&ctx.narrow(&RevealPermutation))
+            .await
+            .unwrap();
+        let mut perms = Vec::new();
+        // TODO(richa) : This is ugly, converted F -> u128 -> usize needed for getting an appropriate Permutation to apply.
+        // Besides this makes the code dependent on Permutation struct needs which is not a good design choice.
+        for i in &permutation {
+            perms.push(i.as_u128() as usize);
+        }
+        let mut permutation = Permutation::oneline(perms);
+
+        // 5. apply the permutation on the shuffled input
+        // The paper suggests taking apply_inv. However, Permutation::oneline returns an inverted permutation, so we are negating
+        // the effect by just calling apply instead of apply_inv here
+        apply(&mut permutation, &mut self.input);
+        Ok(())
+    }
+
+    #[allow(clippy::cast_possible_truncation)]
+    async fn reveal_permutations<N: Network>(
+        &self,
+        ctx: &ProtocolContext<'_, N>,
+    ) -> Result<Vec<F>, BoxError> {
+        let reveals = self
+            .permutation
+            .iter()
+            .enumerate()
+            .map(|(index, input)| async move {
+                reveal(
+                    ctx.narrow(&index.to_string()),
```

Comment:
Not sure if this is correct? I am calling a narrow per record ID for reveal. But removing this gives :"'Refined 'protocol' with step 'reveal_permutation' twice'"

Reply:
Honestly, I can never get this right either. I suggest just land whatever hack works, and we will do a cleanup pass over the entire "step" thing in all the protocols later.

As soon as we have an end-to-end testable protocol and we can profile it, I'm sure we will find a million crazy things that are all wrong with the way we are using "narrow()".

Reply:
PR #177 makes it possible to reuse same ctx and we dont need to use string any more. So, I am rebasing on top of that PR